bpf/lib/fib.h

```diff
@@ -8,134 +8,172 @@
+		__u16 proto = ctx_get_protocol(ctx);
+
+		if (ctx_change_head(ctx, __ETH_HLEN, 0))
+			return DROP_INVALID;
+		if (eth_store_proto(ctx, proto, 0) < 0)
+			return DROP_WRITE_ERROR;
+	}
+	return 0;
+}
+
+static __always_inline bool fib_ok(int ret)
+{
+	return likely(ret == CTX_ACT_TX || ret == CTX_ACT_REDIRECT);
+}
+
 static __always_inline int
-redirect_direct_v6(struct __ctx_buff *ctx __maybe_unused,
-		   int l3_off __maybe_unused,
-		   struct ipv6hdr *ip6 __maybe_unused, int *oif)
+fib_redirect(struct __ctx_buff *ctx, const bool needs_l2_check,
+	     struct bpf_fib_lookup_padded *fib_params, __s8 *fib_err, int *oif)
 {
+	struct bpf_redir_neigh nh_params;
+	struct bpf_redir_neigh *nh = NULL;
 	bool no_neigh = is_defined(ENABLE_SKIP_FIB);
 	int ret;
-	struct bpf_redir_neigh *nh = NULL;
-# ifndef ENABLE_SKIP_FIB
-	struct bpf_redir_neigh nh_params;
-	struct bpf_fib_lookup fib_params = {
-		.family		= AF_INET6,
-		.ifindex	= ctx->ingress_ifindex,
-	};
 
-	ipv6_addr_copy((union v6addr *)&fib_params.ipv6_src,
-		       (union v6addr *)&ip6->saddr);
-	ipv6_addr_copy((union v6addr *)&fib_params.ipv6_dst,
-		       (union v6addr *)&ip6->daddr);
-
-	ret = fib_lookup(ctx, &fib_params, sizeof(fib_params),
+#ifndef ENABLE_SKIP_FIB
+	ret = fib_lookup(ctx, &fib_params->l, sizeof(fib_params->l),
 			 BPF_FIB_LOOKUP_DIRECT);
-	switch (ret) {
-	case BPF_FIB_LKUP_RET_SUCCESS:
-		break;
-	case BPF_FIB_LKUP_RET_NO_NEIGH:
-		nh_params.nh_family = fib_params.family;
-		__bpf_memcpy_builtin(&nh_params.ipv6_nh, &fib_params.ipv6_dst,
-				     sizeof(nh_params.ipv6_nh));
-		no_neigh = true;
-		nh = &nh_params;
-		break;
-	default:
-		return CTX_ACT_DROP;
+	if (ret != BPF_FIB_LKUP_RET_SUCCESS) {
+		*fib_err = (__s8)ret;
+		if (likely(ret == BPF_FIB_LKUP_RET_NO_NEIGH)) {
+			nh_params.nh_family = fib_params->l.family;
+			__bpf_memcpy_builtin(&nh_params.ipv6_nh,
+					     &fib_params->l.ipv6_dst,
+					     sizeof(nh_params.ipv6_nh));
+			nh = &nh_params;
+			no_neigh = true;
+			/* For kernels without d1c362e1dd68 ("bpf: Always
+			 * return target ifindex in bpf_fib_lookup") we
+			 * fall back to use the caller-provided oif when
+			 * necessary.
+			 */
+			if (!is_defined(HAVE_FIB_IFINDEX) && *oif > 0)
+				goto skip_oif;
+		} else {
+			return DROP_NO_FIB;
+		}
 	}
-
-	*oif = fib_params.ifindex;
-# else
+	*oif = fib_params->l.ifindex;
+skip_oif:
+#else
 	*oif = DIRECT_ROUTING_DEV_IFINDEX;
-# endif /* ENABLE_SKIP_FIB */
+	nh_params.nh_family = fib_params->l.family;
+#endif /* ENABLE_SKIP_FIB */
+	if (needs_l2_check) {
+		bool l2_hdr_required = true;
 
-	ret = ipv6_l3(ctx, l3_off, NULL, NULL, METRIC_EGRESS);
-	if (unlikely(ret != CTX_ACT_OK))
-		return ret;
+		ret = maybe_add_l2_hdr(ctx, *oif, &l2_hdr_required);
+		if (ret != 0)
+			return ret;
+		if (!l2_hdr_required)
+			goto out_send;
+	}
 	if (no_neigh) {
-		if (nh)
-			return redirect_neigh(*oif, nh, sizeof(*nh), 0);
-		else
-			return redirect_neigh(*oif, NULL, 0, 0);
+		if (neigh_resolver_available()) {
```

Comment:
I think it would be great to have a comment explaining this path. Also, the `fib_redirect()` function comment would help for future readers to understand this datapath helper better.